parser/parser_test.go

```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+			actual.name,
+			actual.item,
+			expected.name,
+			expected.item))
+	}
+
+	equal, diff = compareObjectsClasses(actual, expected, depth)
+	if !equal {
+		return equal, diff
+	}
+
+	switch actual.item.Kind() {
+	case reflect.Array, reflect.Slice:
+		return deepCompareArrays(actual, expected, depth)
+	case reflect.Map:
+		return deepCompareMaps(actual, expected, depth)
+	case reflect.Struct:
+		return deepCompareStructs(actual, expected, depth)
+	case reflect.Ptr, reflect.Interface:
+		return deepComparePointers(actual, expected, depth)
+	default:
+		return comparePrimitives(actual, expected, depth)
+	}
+
+}
+
+func deepComparePointers(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	if actual.item.IsNil() || expected.item.IsNil() {
+		if actual.item.IsNil() && expected.item.IsNil() {
+			return true, ""
+		} else {
+
+			if actual.item.IsNil() {
+				return false, fmt.Sprintf("%s is nil whereas %s is %#v", actual.name, expected.name, expected.item)
+			} else {
+				return false, fmt.Sprintf("%s is nil whereas %s is %#v", expected.name, actual.name, actual.item)
 			}
+		}
+
+	}
+	return deepCompareObjects(namedItem{actual.name, actual.item.Elem()},
+		namedItem{expected.name, expected.item.Elem()},
+		depth)
+}
+
+func comparePrimitives(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	// TODO make this less hacky
+	var a, b interface{}
+	if actual.item.Kind() != expected.item.Kind() {
+		return false, fmt.Sprintf("%s is of type %s whereas %s is of type %s", actual.name, actual.item.Kind(), expected.name, expected.item.Kind())
+	}
+	switch actual.item.Kind() {
+	case reflect.String:
+		a = actual.item.String()
+		b = expected.item.String()
+	case reflect.Bool:
+		a = actual.item.Bool()
+		b = expected.item.Bool()
+	case reflect.Int:
+		a = actual.item.Int()
+		b = expected.item.Int()
+	default:
+		panic(fmt.Sprintf("unrecognized object types, %s (%#v) and %s (%#v)", actual.name, actual.item, expected.name, expected.item))
+	}
+	if reflect.DeepEqual(a, b) {
+		return true, ""
+	} else {
+		return false, fmt.Sprintf("%s = %#v whereas %s = %#v", actual.name, actual.item, expected.name, expected.item)
+	}
+
+}
+
+func deepCompareMaps(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	a := actual.item
+	b := expected.item
+	aKeys := a.MapKeys()
+	bKeys := b.MapKeys()
+	for _, aKey := range aKeys {
+		if !(isIn(aKey, bKeys)) {
+			aValue := a.MapIndex(aKey)
+			return false, fmt.Sprintf("%v contains key %s (with value %#v) but %v does not", actual.name, aKey, aValue, expected.name)
+		}
+	}
+	for _, bKey := range bKeys {
+		if !(isIn(bKey, aKeys)) {
+			bValue := b.MapIndex(bKey)
+			return false, fmt.Sprintf("%v contains key %s (with value %#v) but %v does not", expected.name, bKey, bValue, actual.name)
+		}
+	}
+	for _, aKey := range aKeys {
+		bKey, _ := findMatching(aKey, bKeys)
```

Comment:
Yeah, I was doing a direct map lookup initially. Since the keys are themselves pointers pointing to different but recursively equal objects, the lookup wasn't finding the match.

Technically there's a detail here that doesn't get tested directly by the test now, which is that the key in the map not only should be recursively equal to another map key in the original object, but also that it should be the same reference as another object in the parse tree somewhere. For the moment that check is in the builder class itself though.

The more that I work on this parser test, the more I wonder if it would be better to have the parser test use the printer and do a text diff instead, and then to reimplement the printer test to directly pass parseTrees into the printer and to not have it use the parser at all. I guess my default is to not bother to rearrange the parser and printer tests, but I could do that too if you think it would be better. Technically that approach wouldn't be able to rule out the possibility that the parser was outputting some unexpected data structure that the printer just so happened to accept and just so happened to print in a way that matched the desired output - I hypothesize we wouldn't be very concerned about that possibility either.